Update documentation/parsers to allow different languages

FabioRosado opened the pull request:
# Description

Start tackling work for #422, since only two of the parsers need the language to be specified in the API Headers the documentation was updated as well to mention how to use the matcher in a different language.

## Status
**READY**


## Type of change

_Please delete options that are not relevant._


- New feature (non-breaking change which adds functionality)
- Documentation (fix or adds documentation)


# How Has This Been Tested?

_Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration._

- Unittest + Tox
- Try to use recastai without a lang parameter in the configuration. 


# Checklist:

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation (if applicable)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

jacobtomlinson commented:
Thanks for raising this!

I see you've just added the `lang` option to the recast.ai config. Will it not get from the main `lang` parameter in the config file? Are there benefits from also configuring it separately? 

I seem to remember this being in @afibanez s example so perhaps he has some thoughts.

afibanez commented:
Yeah, I think that for now it's better to have lang configured one time in main configuration.
Configure the lang inside the parser config only has sense if we move to a multilang implementation (having multiple languages working at same time),
But as we agree, the objective now is localize opsdroid with one language active.

FabioRosado commented:
I see what you mean, I've tried to get the lang parameter from the main config but I wasn't able to do it yet. If I try to add opsdroid to the `call_recastai` opsdroid hangs. I'll try to think for a way to get the main config file and use that lang parameter 👍

jacobtomlinson commented:
You could add it to the connector config in `parse_recastai`.

```python
async def parse_recastai(opsdroid, message, config):
    """Parse a message against all recastai intents."""
    if 'lang' not in config:
        config['lang'] = opsdroid.config.get('lang', 'en')
    ...
```

FabioRosado commented:
This is what I came up with, it seems to work okay. Could you let me know what you think about it?
I've also added the language reference in the `configuration-reference.md`

jacobtomlinson commented on docs/configuration-reference.md:
````diff
@@ -191,6 +191,14 @@ This timezone will be used in crontab skills if the timezone has not been set as
+
+To use opsdroid with a different language other than English you can specify it in your configuration.yaml. The language code needs to be in the standardized [ISO 639-1](https://en.wikipedia.org/wiki/List_of_ISO_639-1_codes).
+```yaml
+language: <ISO 639-1 code -  example: 'en'>
````
This should be `lang` not `language`.

jacobtomlinson replied:
You might also want to mention that it defaults to English.

FabioRosado replied:
I wasn't sure what to use so I thought using the full term would be more specific, i'll change this in the fix

jacobtomlinson commented on docs/matchers/dialogflow.md:
```diff
@@ -30,6 +30,11 @@ The above skill would be called on any intent which has a name of `'myIntent'`.
 
 You can find a quick getting started with Dialogflow guide [here](https://dialogflow.com/docs/getting-started/basics).
 
+If you want to use Dialogflow in a different language other than English, all you need to do is specify the `lang` parameter in Opsdroid's configuration. Then change/add another language to your Dialogflow agent in the Language tab of the agent settings.
```
Let's have a lowercase opsdroid for consistency.

jacobtomlinson commented on docs/matchers/recast.ai.md:
```diff
@@ -36,6 +36,10 @@ You need to [register](https://recast.ai/signup) on Recast.AI and create a bot i
 
 You can find a quick getting started with the Recast.AI guide [here](https://recast.ai/docs/create-your-bot).
 
+If you want to use Recast.AI in a different language other than English, all you need to do is specify the `lang` parameter in Opsdroid's configuration.
```
Lower case opsdroid again.

FabioRosado commented:
Do you agree with the change on the `call_recastai` and `call_dialogflow`? I need to add a default option on the lang param tho.

jacobtomlinson commented:
Yeah I'm happy with that. You're already setting the default option in the `config.get()` call so I wouldn't double up. You might want to create a constant in `const.py` called something like `DEFAULT_LANGUAGE` and import and use that rather than dropping the `'en'` string around.

FabioRosado commented:
I didn't want to touch the tests to make this PR pass so I changed things around a bit. Added the DEFAULT_LANGUAGE int he const and used it as a default to `lang` in the `call_parser` function.

Changed the `.get('lang', 'en')` to just `.get('lang')` if the key is not found the default value is still replaced.

I tried using different lang strings in the config to see if the parser would show the right language (it does), if no value is passed, then the DEFAULT_LANGUAGE value is used.

Let me know what you think 😄

jacobtomlinson commented:
If you do not specify a default option in `.get()` it will return `None`. This will then be passed to the method and so the `DEFAULT_LANGUAGE` will never be used.

FabioRosado commented:
Ah I see. Since the test was showing the English language I assumed everything was going okay. 
Am I wrong thinking that the default value in the lang should be used due to backwards compatibility? Or I can just add the `lang` param in the tests?

jacobtomlinson commented:
Feel free to test both.

You're right that it should default to 'en' for backward compatibility, but at the same time you are also updating every call to that function so you can be pretty sure that you wont break backward compatibility.

I would just restore the default option in the `.get()` methods using the `DEFAULT_LANGUAGE` constant then you've covered all the bases. It's one of the reasons I suggested creating the constant, so you can use it in both places.

FabioRosado commented:
Okay I've added the default value in the .get() and removed it from the call_parser method. 
I had to update the test to add another param to call_parser (Just added None). Do you think this would be ok?

afibanez commented:
I think it's better to maintain the default arg in the method definition, and then use it in the get:
```pyhon
async def call_dialogflow(message, config, lang=DEFAULT_LANGUAGE):
[...]
opsdroid.config.get("lang", lang))
```

jacobtomlinson commented:
@afibanez the point from before is that the `opsdroid` object is not accessible in `call_dialogflow` so what you propose is not possible.

I was suggesting having `DEFAULT_LANGUAGE` in **both** the method definition and the `opsdroid.config.get()` method. It's potentially unnecessary but it covers all bases.

afibanez commented:
@jacobtomlinson ok! I was tricked reading the change in github, with the summarize of changes i was confused about where the get line is :confused:

FabioRosado commented:
Apologies about this, I miss understood what you meant Jacob and kind off messed up a bit.
I've applied the changes to the call_parser and reverted the tests back to normal since we don't need to touch them to pass. Thanks for all the help!
